### PR TITLE
Add script to setup node

### PR DESCRIPTION
**Pull request by DmitriVT**

### CODE REVIEW

**Comment by DmitriVT** [comment]
https://www.notion.so/Run-a-validator-flow-b1c84fd1c2aa4c2caf60e6472a479f63 how to use validator node

File: scripts/run_validator_node.sh [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+            chain_spec="testnet"
+            ;;
+        *)
+            echo "Invalid choice. Please select a right number."
+            continue
+            ;;
+    esac
+done
+
+echo "Your choose is: $chain_spec network."
```
Comment (ales-tsurko) [inline scripts/run_validator_node.sh]: "Your choice"

File: scripts/run_validator_node.sh [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+
+if [ "$(ls -A /opt/atleta/data/chains &>/dev/null)" ]; then
```
Comment (ales-tsurko) [inline scripts/run_validator_node.sh]: this is not a standard path, we should ask the user where the chain data is located

File: scripts/run_validator_node.sh [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+	keychain_exists=1
+else
+	keychain_exists=0
+fi
+
+echo "Everything's ready. Tasks:"
+echo "  [X] Download atleta-node -> /usr/local/bin/atleta-node"
+echo "  [X] Download $chain_spec network chain spec -> /opt/atleta/chain_spec.json"
```
Comment (ales-tsurko) [inline scripts/run_validator_node.sh]: the better place for this is `~/.config/atleta` directory. Also, we will name the chain spec file in accordance to the network (i.e. `chain-spec.testnet.json`, `chain-spec.mainnet.json`, etc.)

File: scripts/run_validator_node.sh [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+fi
+echo "Press Enter to continue or Ctrl-C to cancel."
+read < /dev/tty
+
+echo "Create directories..."
+$sudo_cmd mkdir -p /opt/atleta/ /usr/local/bin/
+echo "Stop old node if it's running..."
+
+#Check status of process
+if [[ $arch == "x86_64" ]]; then
```
Comment (ales-tsurko) [inline scripts/run_validator_node.sh]: condition based only on architecture won't work as linux may be of both architectures. (not sure macos still can be intel, but it's better to check as well). you should rather check whether you are on linux or macos. also, with `systemctl` here we suppose that the user uses `systemd`, so we need to either make it a requirement or modify it so that it would work without `systemd`.

File: scripts/run_validator_node.sh [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+    $sudo_cmd systemctl stop atleta-validator &>/dev/null || true
+else
+    $sudo_cmd launchctl unload /Library/LaunchDaemons/com.example.atleta-validator.plist || true
+fi
+
+echo "Download binary..."
+$sudo_cmd curl -sSL https://github.com/Atleta-network/atleta/releases/download/v1.0.0/atleta-node -o /usr/local/bin/atleta-node
```
Comment (ales-tsurko) [inline scripts/run_validator_node.sh]: you should check whether `curl` is installed on the user's machine first.

File: scripts/run_validator_node.sh [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+$sudo_cmd chmod +x /usr/local/bin/atleta-node
+echo "Download chain spec..."
+$sudo_cmd curl -sSL https://github.com/Atleta-network/atleta/releases/download/v1.0.0/$chain_spec-chain-spec.json -o /opt/atleta/chain_spec.json
+
+echo "Generate atleta-validator.service..."
+
+if [[ $arch == "x86_64" ]]; then
```
Comment (ales-tsurko) [inline scripts/run_validator_node.sh]: here is the same issue with determining the operating system as I described above.

File: scripts/run_validator_node.sh [hunk 7]
```diff
@@ -0,0 +1,187 @@
...
+	while [ -z "$session_keys" ]; do
+		echo "Waiting..."
+		if [ $i -gt 60 ]; then
+			echo
+			echo "Node didn't start after 5 minutes." >&2
+			exit 1
+		fi
+		sleep 5
+		set +e
+		session_keys=$(wscat -c ws://127.0.0.1:9944 -x '{"id":1, "jsonrpc":"2.0", "method": "author_rotateKeys", "params":[]}' 2>/dev/null | jq -r .result)
```
Comment (ales-tsurko) [inline scripts/run_validator_node.sh]: please, check this works. probably, you need to run the node with `--rpc-methods=Unsafe`

File: scripts/run_validator_node.sh (hunk 2 above)
Comment (ales-tsurko) [reply scripts/run_validator_node.sh]: for macos it's better to store it in the user's home directory (in a dedicated directory).

**Review by ales-tsurko — CHANGES_REQUESTED** [review]

File: scripts/run_validator_node.sh (hunk 7 above)
Comment (ales-tsurko) [reply scripts/run_validator_node.sh]: also, use curl. wscat is not so common tool

Comment (ales-tsurko) [reply scripts/run_validator_node.sh]: and you need `jq` to be installed on the user's machine, so check it's installed and return the error with installation instructions if it's not.

File: scripts/run_validator_node.sh (hunk 5 above)
Comment (DmitriVT) [reply scripts/run_validator_node.sh]: Was added to documentation in the "Requirements"

File: scripts/run_validator_node.sh (hunk 7 above)
Comment (DmitriVT) [reply scripts/run_validator_node.sh]: --rpc-methods=Unsafe is set by default. We can rotateKeys without this flag.
Replaced wscat with curl.
Added jq to doc

File: scripts/run_validator_node.sh (hunk 3 above)
Comment (DmitriVT) [reply scripts/run_validator_node.sh]: Agree with you. Done.

File: scripts/run_validator_node.sh [hunk 8]
```diff
@@ -0,0 +1,232 @@
...
+echo "Detecting your architecture... "
+operating_system=$(uname -s)
+if [ "$operating_system" != "Linux" -a "$operating_system" != "Darwin" ]; then
+	echo "$operating_system is currently not supported by Atleta Node, only Linux or Darwin are supported. Exiting." >&2
+	exit 1
+else
+	echo "OK ($operating_system)"
+fi
+
+if [[ $arch == "Linux" ]]; then
```
Comment (ales-tsurko) [inline scripts/run_validator_node.sh]: you confused variable name, please check the script works

File: scripts/run_validator_node.sh [hunk 9]
```diff
@@ -0,0 +1,232 @@
...
+
+#Check status of process
+if [[ $arch == "Linux" ]]; then
+    $sudo_cmd systemctl stop atleta-validator &>/dev/null || true
+else
+    $sudo_cmd launchctl unload /Library/LaunchDaemons/com.example.atleta-validator.plist || true
+fi
+
+echo "Download binary..."
+$sudo_cmd curl -sSL https://github.com/Atleta-network/atleta/releases/download/v1.0.0/atleta-node -o ~/.config/atleta/atleta-node
```
Comment (ales-tsurko) [inline scripts/run_validator_node.sh]: in your previous script the binary file path was actually good, and putting it into ~/.config is a bad idea as it's rarely included in PATH variable.

File: scripts/run_validator_node.sh [hunk 10]
```diff
@@ -0,0 +1,232 @@
...
+$sudo_cmd chmod +x /usr/local/bin/atleta-node
```
Comment (ales-tsurko) [inline scripts/run_validator_node.sh]: yeah, and here you use the old path...

File: scripts/run_validator_node.sh [hunk 11]
```diff
@@ -0,0 +1,232 @@
...
+echo "Download chain spec..."
+$sudo_cmd curl -sSL https://github.com/Atleta-network/atleta/releases/download/v1.0.0/chain_spec.$chain_spec.json -o ~/.config/atleta/chain_spec.$chain_spec.json
```
Comment (ales-tsurko) [inline scripts/run_validator_node.sh]: but the chain spec file path is good to be ~/.config

File: specs/README.md
Comment (ales-tsurko) [inline specs/README.md:1]: please check this directory name to "chainspecs"

**Review by ales-tsurko — CHANGES_REQUESTED** [review]

File: scripts/run_validator_node.sh (hunk 8 above)
Comment (DmitriVT) [reply scripts/run_validator_node.sh]: Hm, sorry, but isn't relevant file version.
Fixed

File: specs/README.md
Comment (DmitriVT) [reply specs/README.md:1]: Renamed this directory to chainspecs

File: scripts/run_validator_node.sh [hunk 12]
```diff
@@ -0,0 +1,238 @@
+#!/bin/bash
```
Comment (hrls) [inline scripts/run_validator_node.sh]: `#!/usr/bin/env bash` to avoid using the wrong bash on a system with with multiple installed bashes

File: scripts/run_validator_node.sh [hunk 13]
```diff
@@ -0,0 +1,238 @@
...
+
+set -ue
+set -o pipefail
+
+sudo_cmd=""
+files_path=""
```
Comment (hrls) [inline scripts/run_validator_node.sh]: What kind of files? Variable name is too general.

File: scripts/run_validator_node.sh [hunk 14]
```diff
@@ -0,0 +1,238 @@
...
+chain_spec_name=""
+keychain_exists=""
+session_keys=""
+binary_path=""
+
+echo "This script will setup a Atleta Validator on your PC. Press Ctrl-C at any time to cancel."
+echo "Checking privileges... "
+if [ $(id -u) -eq 0 ]; then
```
Comment (hrls) [inline scripts/run_validator_node.sh]: ShellCheck says 'Quote this to prevent word splitting'

File: scripts/run_validator_node.sh [hunk 15]
```diff
@@ -0,0 +1,238 @@
...
+    sudo_cmd=sudo
+else
+    echo "FAIL"
+    echo "Not running as root and no sudo detected. Please run this as root or configure sudo. Exiting." >&2
+    exit 1
+fi
+
+echo "Detecting your architecture... "
+operating_system=$(uname -s)
+if [ "$operating_system" != "Linux" -a "$operating_system" != "Darwin" ]; then
```
Comment (hrls) [inline scripts/run_validator_node.sh]: Use `&&` instead `-a`. ShellCheck also beeps here.

File: scripts/run_validator_node.sh [hunk 16]
```diff
@@ -0,0 +1,238 @@
...
+elif command -v sudo &> /dev/null; then
+    echo "OK (sudo)"
+    sudo_cmd=sudo
+else
+    echo "FAIL"
+    echo "Not running as root and no sudo detected. Please run this as root or configure sudo. Exiting." >&2
+    exit 1
+fi
+
+echo "Detecting your architecture... "
```
Comment (hrls) [inline scripts/run_validator_node.sh]: Probably detecting the system arch is more important step than checking privileges. Also it can be useful to use arch variable to detect some features like sudo or root path.

File: scripts/run_validator_node.sh [hunk 17]
```diff
@@ -0,0 +1,238 @@
...
+    if [ -e /run/systemd/system ]; then
+        echo "OK"
+    else
+        echo "FAIL"
+        echo "No systemd detected. Exiting." >&2
+        exit 1
+    fi
+fi
+
+if ! command -v jq &>/dev/null || \
```
Comment (hrls) [inline scripts/run_validator_node.sh]: Split this into two check. Then it will be possible to point what exact requirement is missing. Also in the future another requirement can comes into scope.

File: scripts/run_validator_node.sh [hunk 18]
```diff
@@ -0,0 +1,238 @@
...
+
+echo "Your choice is: $chain_spec_name network."
+
+while [ -z "$files_path" ]; do
+    read -p "Enter the path where you want to store at least a few gigabytes of data (or press Enter to use the standard $HOME/atleta/chain directory): " files_path
+
+    if [ -z "$files_path" ]; then
+        files_path="$HOME/atleta/chain"
+        binary_path="/opt/atleta"
+        $sudo_cmd mkdir -p "$files_path"
```
Comment (hrls) [inline scripts/run_validator_node.sh]: Using sudo is unnecessary to make changes in `$HOME` directory

File: scripts/run_validator_node.sh [hunk 19]
```diff
@@ -0,0 +1,238 @@
...
+done
+
+echo "Your choice is: $chain_spec_name network."
+
+while [ -z "$files_path" ]; do
+    read -p "Enter the path where you want to store at least a few gigabytes of data (or press Enter to use the standard $HOME/atleta/chain directory): " files_path
+
+    if [ -z "$files_path" ]; then
+        files_path="$HOME/atleta/chain"
+        binary_path="/opt/atleta"
```
Comment (hrls) [inline scripts/run_validator_node.sh]: Why binary path differs and depends on what _files_ path we selected?

File: scripts/run_validator_node.sh [hunk 20]
```diff
@@ -0,0 +1,238 @@
...
+        $sudo_cmd mkdir -p "$files_path"
+        $sudo_cmd chmod -R 777 "$files_path"
+        echo "Standard directory selected: $files_path"
+    else
+        files_path="$HOME/$files_path"
+        binary_path="/usr/local"
+        $sudo_cmd mkdir -p "$files_path"
+        $sudo_cmd chmod -R 777 "$files_path"
```
Comment (hrls) [inline scripts/run_validator_node.sh]: `chmod -R a=rwx` symbolic notation is more readable form

File: scripts/run_validator_node.sh [hunk 21]
```diff
@@ -0,0 +1,238 @@
...
+
+    if [ -w "$files_path" ]; then
+        echo "The directory exists and you have write permissions."
+    else
+        echo "You do not have write permission to the specified directory. Please select another directory."
+        files_path=""
+    fi
+done
+
+if [ "$(ls -A $files_path/chains &>/dev/null)" ]; then
```
Comment (hrls) [inline scripts/run_validator_node.sh]: Folder may contain some files that are not related to the chain specs. Like NOTE files created by user or hidden files created by system (e.g. `.DS_Store` on macOS).

File: scripts/run_validator_node.sh [hunk 22]
```diff
@@ -0,0 +1,238 @@
...
+
+#Check status of process
+if [[ $operating_system == "Linux" ]]; then
+    $sudo_cmd systemctl stop atleta-validator &>/dev/null || true
+else
+    $sudo_cmd launchctl unload /Library/LaunchDaemons/com.example.atleta-validator.plist || true
+fi
+
+echo "Download binary..."
+$sudo_cmd curl -sSL https://github.com/Atleta-network/atleta/releases/download/v1.0.0/atleta-node -o "$binary_path/bin/atleta-node"
```
Comment (hrls) [inline scripts/run_validator_node.sh]: These hardcoded URLs should be part of script header for easy editing and overview

File: scripts/run_validator_node.sh [hunk 23]
```diff
@@ -0,0 +1,238 @@
...
+        <key>KeepAlive</key>
+        <dict>
+            <key>SuccessfulExit</key>
+            <false/>
+        </dict>
+    </dict>
+    </plist>
+EOF
+$sudo_cmd launchctl load /Library/LaunchDaemons/com.example.atleta-validator.plist
+$sudo_cmd launchctl start com.example.atleta-validator
```
Comment (hrls) [inline scripts/run_validator_node.sh]: Maybe we can use some real domain here

**Review by hrls — DISMISSED** [review]

**Comment by hrls** [comment]
We should consider using some linter for shell scripts. For example [ShellCheck](https://github.com/koalaman/shellcheck)

File: scripts/run_validator_node.sh (hunk 22 above)
Comment (ales-tsurko) [reply scripts/run_validator_node.sh]: Oh, and just noticed you hard-coded the version. So in case we released 1.0.1, we would need to update the script as well

**Comment by ales-tsurko** [comment]
> We should consider using some linter for shell scripts. For example [ShellCheck](https://github.com/koalaman/shellcheck)

Could you please open a separate issue for this?

**Comment by DmitriVT** [comment]
Added new [issue](https://github.com/orgs/Atleta-network/projects/1/views/1?pane=issue&itemId=62704898) to integrate linter

File: scripts/run_validator_node.sh (hunk 13 above)
Comment (DmitriVT) [reply scripts/run_validator_node.sh]: "files_path" specifies the directory where all the node's data is stored. Maybe it makes sense to rename it "base_path" like flag where can we use it (--base-path <PATH>)

File: scripts/run_validator_node.sh (hunk 18 above)
Comment (DmitriVT) [reply scripts/run_validator_node.sh]: files_path to etc

File: scripts/run_validator_node.sh (hunk 19 above)
Comment (DmitriVT) [reply scripts/run_validator_node.sh]: check operation system files

File: scripts/run_validator_node.sh (hunk 21 above)
Comment (DmitriVT) [reply scripts/run_validator_node.sh]: Will be added to improvements
